test/composableStableJoin.integration.test.ts

```diff
@@ -0,0 +1,387 @@
+// pnpm test -- composableStableJoin.integration.test.ts
+import { describe, expect, test, beforeAll, beforeEach } from 'vitest';
+import { config } from 'dotenv';
+config();
+
+import {
+    Client,
+    createTestClient,
+    http,
+    parseUnits,
+    publicActions,
+    PublicActions,
+    TestActions,
+    WalletActions,
+    walletActions,
+} from 'viem';
+
+import {
+    UnbalancedJoinInput,
+    ProportionalJoinInput,
+    SingleAssetJoinInput,
+    JoinKind,
+    Slippage,
+    Token,
+    TokenAmount,
+    replaceWrapped,
+    Address,
+    Hex,
+    PoolStateInput,
+    CHAINS,
+    ChainId,
+    getPoolAddress,
+    PoolJoin,
+    JoinInput,
+} from '../src';
+import { forkSetup, sendTransactionGetBalances } from './lib/utils/helper';
+
+type TxInput = {
+    client: Client & PublicActions & TestActions & WalletActions;
+    poolJoin: PoolJoin;
+    joinInput: JoinInput;
+    slippage: Slippage;
+    poolInput: PoolStateInput;
+    testAddress: Address;
+    checkNativeBalance: boolean;
+};
+
+const chainId = ChainId.MAINNET;
+const rpcUrl = 'http://127.0.0.1:8545/';
+const blockNumber = BigInt(18043296);
+const poolId =
+    '0x156c02f3f7fef64a3a9d80ccf7085f23cce91d76000000000000000000000570'; // Balancer vETH/WETH StablePool
+
+describe('composable stable join test', () => {
+    let txInput: TxInput;
+    let bptToken: Token;
+
+    beforeAll(async () => {
+        // setup mock api
+        const api = new MockApi();
+
+        // get pool state from api
+        const poolInput = await api.getPool(poolId);
+
+        const client = createTestClient({
+            mode: 'hardhat',
+            chain: CHAINS[chainId],
+            transport: http(rpcUrl),
+        })
+            .extend(publicActions)
+            .extend(walletActions);
+
+        txInput = {
+            client,
+            poolJoin: new PoolJoin(),
+            slippage: Slippage.fromPercentage('1'), // 1%
+            poolInput,
+            testAddress: '0x10a19e7ee7d7f8a52822f6817de8ea18204f2e4f', // Balancer DAO Multisig
+            joinInput: {} as JoinInput,
+            checkNativeBalance: false,
+        };
+
+        // setup BPT token
+        bptToken = new Token(chainId, poolInput.address, 18, 'BPT');
+    });
+
+    beforeEach(async () => {
+        await forkSetup(
+            txInput.client,
+            txInput.testAddress,
+            [...txInput.poolInput.tokens.map((t) => t.address)],
+            [0, 0, 3],
+            [
+                ...txInput.poolInput.tokens.map((t) =>
+                    parseUnits('100', t.decimals),
+                ),
+            ],
+            process.env.ETHEREUM_RPC_URL as string,
+            blockNumber,
+        );
+    });
+
+    test('unbalanced join', async () => {
```

Comment:
Maybe an opportunity to reuse code could come from identifying patterns between `unbalanced join` tests on different types of pools (e.g. weighted, composable stable, etc..).
I feel like the assertions should be pretty similar between them. If that's true, we could have something like a shared `testUnbalancedJoin` between pool types.
Likely not something for this PR, but for a separate one.

Reply:
Created an Asana for this.